### DISCUSSION TITLE
Performance Improvements

### DISCUSSION THREAD

**Discussion by not-matthias** [solidtime-io/solidtime#260] [Feature requests]
SolidTime is really nice, but it "feels" quite slow. Looking at the network tab, I can see that it's downloading 10s of KB of data (dashboard, time entries, ...) very often. For comparison, SuperProductivity (which also uses Electron/Web) feels super fast and smooth. 

Possible solution: **Offline-first** application, that is **synced** with the server.

**Comment by Onatcer** [comment]
are you talking about the web app or the desktop app? and are there specific examples where it feels slow? 

generally speaking, we aim for a local-first approach utilizing tanstack query in the desktop app, but we are not fully there yet. The data management in the web app is different, and will only update after the server request succeeds. The long-term goal for the web app is to have a local-first approach there as well. 

One of the reasons why you might see a lot of API Requests is that we refresh all data when the window is focused. In the Web app this might also happen if you open the dev tools, not sure about that tho. We do this to combat problems with desync between multiple clients (f.e. when you use both desktop and web app). 

Edit: I see switching to the Dashboard seems to be quite slow for me right now, we'll look into that. Generally for navigation we have a delay because we currently use Inertia which will always send a request before it navigates to a new page. The plan is to move away from Inertia to a pure SPA in the future and most of the application is already built with that in mind.

**Reply by not-matthias** [reply]
Thanks for the response and for looking into it!

> are you talking about the web app or the desktop app? and are there specific examples where it feels slow?

I was using the Website. Examples include going from Dashboard to Time/Reporting. Also clicking on the 'Start Timer' button has a delay. 

> One of the reasons why you might see a lot of API Requests is that we refresh all data when the window is focused. In the Web app this might also happen if you open the dev tools, not sure about that tho. We do this to combat problems with desync between multiple clients (f.e. when you use both desktop and web app).

A more lightweight approach to this would probably be a Websocket channel where you can send/receive updates from other clients. I've seen many websites use this approach already. 

> Generally for navigation we have a delay because we currently use Inertia which will always send a request before it navigates to a new page. The plan is to move away from Inertia to a pure SPA in the future and most of the application is already built with that in mind.

Having a delay seems weird :sweat_smile: But I'm sure it has it's right. Glad to hear that you are planning to move to pure SPA! Can't wait haha. Thanks for all the hard work!

**Reply by korridor** [reply]
@not-matthias The slow dashboard page problem should be fixed now in the `main` tag and on solidtime Cloud.

**Reply by enieuwy** [reply]
I'm also finding it very slow (v0.8.0). No obvious reason why, and it wasn't always this slow.

**Reply by Onatcer** [reply]
@enieuwy is there anything specific that is slow? this discussion is a few months old, we addressed a few of the issues, but I cannot think of any changes we made recently that should impact the performance negatively.
